merlin/models/torch/blocks/experts.py

```diff
@@ -0,0 +1,244 @@
+            "num_task_experts": num_task_experts,
+            "outputs": outputs,
+        }
+        super().__init__(*CGCBlock(shared_gate=True, **cgc_kwargs).repeat(depth - 1))
+        self.append(CGCBlock(**cgc_kwargs))
+
+
+class CGCBlock(Block):
+    """
+    Implements the Customized Gate Control (CGC) proposed in [1].
+
+    {ple_reference}
+
+    Parameters
+    ----------
+    expert : nn.Module
+        The base expert model to be used.
+    num_shared_experts : int
+        The number of shared experts.
+    num_task_experts : int
+        The number of task-specific experts.
+    outputs : ParallelBlock
+        The output block.
+    shared_gate : bool, optional
+        If true, use a shared gate for all tasks. Defaults to False.
+    """
+
+    def __init__(
+        self,
+        expert: nn.Module,
+        *,
+        num_shared_experts: int,
+        num_task_experts: int,
+        outputs: ParallelBlock,
+        shared_gate: bool = False,
+    ):
+        shared_experts = repeat_parallel(expert, num_shared_experts, agg=Stack(dim=1))
+        expert_shortcut = partial(ShortcutBlock, output_name="experts")
+        super().__init__(expert_shortcut(shared_experts))
+
+        gates = ParallelBlock()
+        for name in outputs.branches:
+            gates.branches[name] = PLEExpertGateBlock(
+                num_shared_experts + num_task_experts,
+                experts=repeat_parallel(expert, num_task_experts, agg=Stack(dim=1)),
+                name=name,
+            )
+        if shared_gate:
+            gates.branches["experts"] = expert_shortcut(
+                ExpertGateBlock(num_shared_experts), propagate_shortcut=True
+            )
+
+        self.append(gates)
+
+
+class ExpertGateBlock(Block):
+    """Expert Gate Block.
+
+    # TODO: Add initialize_from_schema to remove the need to pass in num_experts
+
+    Parameters
+    ----------
+    num_experts : int
+        The number of experts used.
+    """
+
+    def __init__(self, num_experts: int):
+        super().__init__(GateBlock(num_experts))
+
+    def forward(
+        self, inputs: Union[torch.Tensor, Dict[str, torch.Tensor]], batch: Optional[Batch] = None
+    ) -> torch.Tensor:
+        if torch.jit.isinstance(inputs, torch.Tensor):
+            raise RuntimeError("ExpertGateBlock requires a dictionary input")
+
+        experts = inputs["experts"]
+        outputs = inputs["shortcut"]
+        for module in self.values:
+            outputs = module(outputs, batch=batch)
+
+        # return torch.sum(experts * outputs, dim=1, keepdim=False)
+        gated = outputs.expand_as(experts)
+
+        # Multiply and sum along the experts dimension
+        return (experts * gated).sum(dim=1)
+
+
+class PLEExpertGateBlock(Block):
+    """
+    Progressive Layered Extraction (PLE) Expert Gate Block.
+
+    Parameters
+    ----------
+    num_experts : int
+        The number of experts used.
+    experts : nn.Module
```

Comment:
Could this `experts` argument maybe be renamed to `task_experts`, to make it clear they are different from the shared experts whose output is fed within the gate?